### PR TITLE
Darc command to add goal

### PR DESCRIPTION
**Pull request by epananth**
1) Updated Maestro client code generation 
2) Added darc goal commands for adding/ updating and fetching goal data. 
3) Resolved some compile errors in the process.
4) added the generated code

### CODE REVIEW

File: src/Maestro/Client/src/ApiFactory.cs [hunk 1]
```diff
@@ -3,30 +3,61 @@
...
 using System;
-using Microsoft.Rest;
 
 namespace Microsoft.DotNet.Maestro.Client
 {
     public static class ApiFactory
     {
+        /// <summary>
+        /// Obtains API client for authenticated access to internal queues.
+        /// The client will access production Helix instance.
```
Comment (alexperovich) [inline src/Maestro/Client/src/ApiFactory.cs]: These comments need to be fixed. They currently refer to Helix.

File: src/Maestro/Client/src/MaestroApi.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        public override bool IsRetriableException(Exception exception)
+        {
+            return base.IsRetriableException(exception) ||
+                   exception is TaskCanceledException ||
+                   exception is OperationCanceledException ||
+                   exception is HttpRequestException ||
+                   exception is RestApiException raex && raex.Response.Status >= 500 && raex.Response.Status <= 599 ||
+                   exception is IOException ||
+                   exception is SocketException ||
+                   exception is RestApiException jobListEx && jobListEx.Response.Status == 400 && jobListEx.Message.Contains("Provided Job List Uri is not accessible");
```
Comment (alexperovich) [inline src/Maestro/Client/src/MaestroApi.cs]: This condition is specific to the Helix Api. It can be removed for maestro.

File: src/Microsoft.DotNet.Darc/src/DarcLib/MaestroApiBarClient.cs [hunk 3]
```diff
@@ -254,7 +254,7 @@ public async Task<IEnumerable<MergePolicy>> GetRepositoryMergePoliciesAsync(stri
             {
                 return await _barClient.Repository.GetMergePoliciesAsync(repository: repoUri, branch: branch);
             }
-            catch (RestApiException e) when (e.Response.StatusCode == HttpStatusCode.NotFound)
+            catch (RestApiException e) when (e.Response.Status.Equals(HttpStatusCode.NotFound))
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Darc/src/DarcLib/MaestroApiBarClient.cs]: I don't think this will work. Use the same syntax as the other checks `Status == (int) <code>`

Comment (epananth) [reply src/Microsoft.DotNet.Darc/src/DarcLib/MaestroApiBarClient.cs]: I changed this, may be I missed this one place.

File: src/Maestro/Client/src/MaestroApi.cs (hunk 2 above)
Comment (epananth) [reply src/Maestro/Client/src/MaestroApi.cs]: okay removing condition and keeping the override method only.

File: src/Maestro/Client/src/Microsoft.DotNet.Maestro.Client.csproj [hunk 4]
```diff
@@ -12,7 +12,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Rest.ClientRuntime" Version="$(MicrosoftRestClientRuntimeVersion)" />
+    <PackageReference Include="Azure.Core" Version="1.0.1" />
```
Comment (epananth) [inline src/Maestro/Client/src/Microsoft.DotNet.Maestro.Client.csproj]: @alexperovich we made this change in my local, is it okay to tie it down to this particular version?

**Comment by ChadNedzlek** [comment]
I sort of wish we'd split the "radically change the underlying API wrapper" from other functional changes, because it's hard to pull the two apart to code review them. :-(

File: src/Maestro/Client/src/MaestroApi.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+using System.Net.Http;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.DotNet.Maestro.Client
+{
+    partial class MaestroApiResponseClassifier
+    {
+        public override bool IsRetriableException(Exception exception)
```
Comment (ChadNedzlek) [inline src/Maestro/Client/src/MaestroApi.cs]: This isn't a useful class...

File: src/Maestro/Client/src/MaestroApi.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            return base.IsRetriableException(exception);
+        }
+    }
+
+    partial class MaestroApi
+    {
+        partial void HandleFailedRequest(RestApiException ex)
+        {
+            if (ex.Response.Status == (int)HttpStatusCode.BadRequest)
```
Comment (ChadNedzlek) [inline src/Maestro/Client/src/MaestroApi.cs]: What is this method supposed to do? There are a lot of silent returns in here. That seems bad.

File: src/Maestro/Client/src/MaestroApiOptions.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+using Azure.Core;
+using Azure.Core.Pipeline;
+using System;
+
+namespace Microsoft.DotNet.Maestro.Client
+{
+    partial class MaestroApiOptions
+    {
+        private MaestroApiTokenCredential maestroApiTokenCredential;
```
Comment (ChadNedzlek) [inline src/Maestro/Client/src/MaestroApiOptions.cs]: This doesn't look autogenerated, so it should follow the style guidelines.

File: src/Maestro/Client/src/MaestroApiTokenCredential.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
```
Comment (ChadNedzlek) [inline src/Maestro/Client/src/MaestroApiTokenCredential.cs]: In the previous file, it was "Azure" then "Sytem", in this file, it's the reverse.  We should probably pick one order and stick with it, since these are both new files.

**Comment by epananth** [comment]
@ChadNedzlek I was half way done with my darc command line changes, I wanted to test this locally that's when Alex told me that the Maestro Api changed. So I made those related changes with Alex' help. So I have both the changes clubbed together

File: src/Maestro/Client/src/MaestroApi.cs (hunk 5 above)
Comment (epananth) [reply src/Maestro/Client/src/MaestroApi.cs]: I kind of deleted the other conditions by mistake. Readding the missed conditions.

**Comment by alexperovich** [comment]
@ChadNedzlek unfortunatly the api generator version flows with dependency flow so this repo got the new version. We could downgrade the version to what it was the last time this was generated, but I don't know what that version would be.

File: src/Maestro/Client/src/MaestroApi.cs (hunk 6 above)
Comment (epananth) [reply src/Maestro/Client/src/MaestroApi.cs]: this method is trying to implement the methods from the MaestroApi generated code.

File: src/Maestro/Client/src/MaestroApiTokenCredential.cs (hunk 8 above)
Comment (epananth) [reply src/Maestro/Client/src/MaestroApiTokenCredential.cs]: Rearranged imports.

File: src/Maestro/Client/src/MaestroApi.cs (hunk 6 above)
Comment (alexperovich) [reply src/Maestro/Client/src/MaestroApi.cs]: This method is intended to be any special error handling the consumer of the generated code can do. If this method returns without throwing a specific exception then the generic RestApiException is thrown.

**Comment by ChadNedzlek** [comment]
@alexperovich @epananth   It's probably too late now, but in the future, it would have been nicer to make a new branch to do the update and commit that.  It would have been an easy peasy simple rubber stamp code review.

Then this one would have been a lot cleaner and easier to find the interesting bits.

Though I'll admit, I do the same thing a lot too. It's so... tempting to just keep soldiering forward with an every increasing pile of commits trying to make something work. :-)

File: src/Maestro/Client/src/MaestroApi.cs (hunk 6 above)
Comment (ChadNedzlek) [reply src/Maestro/Client/src/MaestroApi.cs]: Ah. So it has build in handling... Makes sense with the way partials work.  So the early returns are normal.  I wonder if a comment in here somewhere might help future code lookers not get concerned like I did. :-)

Comment (alexperovich) [reply src/Maestro/Client/src/MaestroApi.cs]: Yea, each generated api method has calls to 3 partial methods in the failure path, one HandleFailed{MethodName}Request(), one HandleFailedRequest() on the specific api class, and then one HandleFailedRequest() on the main api class. These allow specific error handling for any possible case.

What method name would you suggest? I couldn't think of anything that made sense.

Comment (ChadNedzlek) [reply src/Maestro/Client/src/MaestroApi.cs]: Yeah... it's a sort of... unusual concept.  I don't have a good verb on the tip of my tounge.  PreHandleFailedRequest? OptionalHandeFailedRequest?  None of them is really better than what we have.  I'd probably just plop some comments in this implementation saying "we don't have anything better to do, return to the caller and let them throw a generic exception".

File: src/Maestro/tests/Scenarios/goal.ps1 [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    # Add a new channel
+    Write-Host "Creating channel '$newChannelName'"
+    Darc-Add-Channel -channelName $newChannelName -classification 'test'
+
+    # Add a new Goal for a Definition in a Channel 
+    Write-Host "Creating goal" 
+    Darc-Set-Goal -channel $newChannelName -definitionId $newDefinitionId -minutes $newMinutes
+
+    # Get the goal and make sure it's there
+    Write-Host "Checking that the goal is added"
```
Comment (adiaaida) [inline src/Maestro/tests/Scenarios/goal.ps1]: Can we update the output messages to match how we do it in the other scripts? So, here:

```
Write-Host "Checking that goal $newMinutes for '$newDefinitionId' on channel '$newChannelName' exists"
```

and below:

```
throw "Cannot find goal $newMinutes for definition '$newDefinitionId' on channel '$newChannelName'"
```

and then also be more descriptive on the else

File: src/Maestro/tests/Scenarios/goal.ps1 [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    # Update goal verification
+    Write-Host "Checking that the goal is getting updated"
+    $goal = Darc-Get-Goal -channel $newChannelName -definitionId $newDefinitionId
+    if (-not $goal -match $newMinutes) {
+        throw "Updating goal failed"
+    } else {
+        Write-Host "Update goal was successful"
+    }
+    Write-Host "Tests passed."
+}finally{
```
Comment (adiaaida) [inline src/Maestro/tests/Scenarios/goal.ps1]:
```suggestion
} finally {
```

File: src/Maestro/tests/Scenarios/goal.ps1 [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    #Teardown only deletes the Channel that was created.
```
Comment (adiaaida) [inline src/Maestro/tests/Scenarios/goal.ps1]: Comment is unnecessary.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/AddChannelOperation.cs [hunk 12]
```diff
@@ -46,12 +46,12 @@ public override async Task<int> ExecuteAsync()
 
                 return Constants.SuccessCode;
             }
-            catch (RestApiException e) when (e.Response.StatusCode == HttpStatusCode.Conflict)
+            catch (RestApiException e) when (e.Response.Status == (int) HttpStatusCode.Conflict)
```
Comment (adiaaida) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/AddChannelOperation.cs]: Why are we making these changes? Seems unrelated to the code you're adding to Darc.

**Comment by adiaaida** [comment]
Other than a few nits on the test, and some confusion around some of the darc related StatusCode -> Status changes, the darc portion of this looks good to me.

I'll leave the API changes to Chad and Alex.

**Comment by alexperovich** [comment]
The StatusCode -> Status changes are required because the code generator has switched to a different runtime library.

**Comment by adiaaida** [comment]
I would like us to really consider splitting the two parts of this change (the API change and Epsi's goal darc command) into two different PRs. I understand that the API change was necessary for testing locally, but I think that it would make a lot more sense if we took a pause, separated the two changes, and made 2 different PRs, first for the API change, and then for the goal change. Since there don't appear to be any files that were changed in common between the two sets of changes, I don't think that this would be too much effort to split apart, and it will make it easier to back out a change if something goes wrong (ie, we won't lose one half of this change if there is a problem in the other half).

File: src/Maestro/tests/Scenarios/goal.ps1 (hunk 11 above)
Comment (epananth) [reply src/Maestro/tests/Scenarios/goal.ps1]: Removed comment

File: src/Maestro/tests/Scenarios/goal.ps1 (hunk 9 above)
Comment (epananth) [reply src/Maestro/tests/Scenarios/goal.ps1]: sure updated logs and exception msgs,

File: src/Maestro/Client/src/MaestroApi.cs (hunk 6 above)
Comment (epananth) [reply src/Maestro/Client/src/MaestroApi.cs]: added comments to address this confusion.

**Comment by epananth** [comment]
@adiaaida I agree that this should have been separated. I had already done a lot of changes and I was trying to get my local working. At that point this seemed the right thing to do. I will definitely keep in mind the next time to segregate changes in case same situation arises. That being said this is one of a kind scenario. Normally these API work locally and due to a hotfix in HelixApi due to process leakage error these generated code differs. So I think we have should allow this PR as is. Since I have written test to check if my functionality works, it is a calculated risk than redoing the entire thing all over again.

File: src/Maestro/Client/src/MaestroApi.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+
+namespace Microsoft.DotNet.Maestro.Client
+{
+    partial class MaestroApiResponseClassifier
+    {
+        public override bool IsRetriableException(Exception exception)
+        {
+            return base.IsRetriableException(exception) ||
+                exception is TaskCanceledException ||
```
Comment (ChadNedzlek) [inline src/Maestro/Client/src/MaestroApi.cs]: TaskCanceledException IS an OperationCanceledException, so line 21 is redundant. :-)

File: src/Maestro/Client/src/MaestroApi.cs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+                JObject content;
+                try
+                {
+                    content = JObject.Parse(ex.Response.Content);
+                }
+                catch (Exception)
+                {
+                    return;
+                }
+                if (content["Message"] is JValue value && value.Type == JTokenType.String)
```
Comment (ChadNedzlek) [inline src/Maestro/Client/src/MaestroApi.cs]: I'd probably just put this in side the try and do
```chsarp
string message = content.Value<string>("Message");
throw new ArgumentException(message, ex):
```
If that doesn't work, the existing catch Exception will do the right thing, and we can simply this line a bunch. (I might even inline the "message" value into the throw for extra simplicity.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+using System.Threading.Tasks;
+using System.Linq;
+using Microsoft.DotNet.Maestro.Client.Models;
+using Microsoft.DotNet.Maestro.Client;
+
+namespace Microsoft.DotNet.Darc.Operations
+{
+    internal class GetGoalOperation : Operation
+    {
+        GetGoalCommandLineOptions _options;
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs]: the "private" needs to be explicit; see the style guidelines

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+        public GetGoalOperation(GetGoalCommandLineOptions options)
+            : base(options)
+        {
+            _options = options;
+        }
+        public override async Task<int> ExecuteAsync()
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs]: There should be a blank line here

File: src/Maestro/Client/src/MaestroApi.cs (hunk 14 above)
Comment (epananth) [reply src/Maestro/Client/src/MaestroApi.cs]: I added this in a try block and looks like it works.

File: src/Maestro/Client/src/MaestroApiOptions.cs (hunk 7 above)
Comment (ChadNedzlek) [reply src/Maestro/Client/src/MaestroApiOptions.cs]: This isn't resolved yet.  The field should be `_maestroApiTokenCredential`, which means the `this` on line 17 can (and should) go away.

File: src/Maestro/Client/src/MaestroApiOptions.cs [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+        public MaestroApiOptions(MaestroApiTokenCredential maestroApiTokenCredential)
+        {
+            this.maestroApiTokenCredential = maestroApiTokenCredential;
+        }
+        partial void InitializeOptions()
+        {
+            if (Credentials != null)
+            {
+                AddPolicy(new BearerTokenAuthenticationPolicy
+                    (Credentials, Array.Empty<string>()), HttpPipelinePosition.PerCall);
```
Comment (ChadNedzlek) [inline src/Maestro/Client/src/MaestroApiOptions.cs]: This is an... unusual wrapping style. I don't think I've ever seen the opening "(" not be attached to the method call/constructor that it's part of.  I'd personally write  it like:
```csharp
  AddPolicy(
    new BearerTokenAuthenticationPolicy(Credentials, Array.Empty<string>()),
    HttpPipelinePositition.PerCall
  );
```

File: src/Maestro/Client/src/MaestroApiTokenCredential.cs [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Microsoft.DotNet.Maestro.Client
+{
+    public class MaestroApiTokenCredential : TokenCredential
+    {
+        public MaestroApiTokenCredential(string token)
+        {
+            Token = token;
+        }
+        public string Token { get; }
```
Comment (ChadNedzlek) [inline src/Maestro/Client/src/MaestroApiTokenCredential.cs]: There needs to be a blank line between all these members

File: src/Maestro/SubscriptionActorService/MaestroBarClient.cs [hunk 19]
```diff
@@ -237,5 +237,28 @@ public Task<Channel> GetChannelAsync(int channelId)
         {
             throw new NotImplementedException();
         }
+
+        /// <summary>
```
Comment (ChadNedzlek) [inline src/Maestro/SubscriptionActorService/MaestroBarClient.cs]: These comments are not true. This method doesn't do any of what these comments say it does. They, in fact, just throw.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+using System.Linq;
+using Microsoft.DotNet.Maestro.Client.Models;
+using Microsoft.DotNet.Maestro.Client;
+
+namespace Microsoft.DotNet.Darc.Operations
+{
+    internal class GetGoalOperation : Operation
+    {
+        private GetGoalCommandLineOptions _options;
+        public GetGoalOperation(GetGoalCommandLineOptions options)
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs]: Needs a blank line

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+        /// Retrieve Goal in minutes for Definition in a Channel.
+        /// </summary>
+        /// <returns>Process exit code.</returns>
+        public override async Task<int> ExecuteAsync()
+        {
+            try
+            {
+                IRemote remote = RemoteFactory.GetBarOnlyRemote(_options, Logger);
+                Goal goalInfo = await remote.GetGoalAsync(_options.Channel, _options.DefinitionId);
+                if(goalInfo != null)
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs]: This is an... odd check. If it _is_ null we just... do nothing? Does that make any sense?  We should probably either just assume it's null and not check (since we own the other side of this pipe, and null isn't a valid answer), or have some sort of answer for if it's null.  Silently do nothing but also return success is very unexpected behavior.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/SetGoalOperations.cs [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+        ///  Sets Goal in minutes for a definition in a Channel.
+        /// </summary>
+        /// <returns>Process exit code.</returns>
+        public override async Task<int> ExecuteAsync()
+        {
+            try
+            {
+                IRemote remote = RemoteFactory.GetBarOnlyRemote(_options, Logger);
+                Goal goalInfo = await remote.SetGoalAsync(_options.Channel, _options.DefinitionId, _options.Minutes);
+                return Constants.SuccessCode;
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/SetGoalOperations.cs]: Should we output something about "goal set" or something, maybe outputting some of the values from the "goalInfo" object for the users information?

File: src/Microsoft.DotNet.Darc/src/DarcLib/MaestroApiBarClient.cs [hunk 23]
```diff
@@ -401,5 +401,30 @@ public Task<Build> UpdateBuildAsync(int buildId, BuildUpdate buildUpdate)
...
+        ///     Creates a new goal or updates the existing goal (in minutes) for a Defintion in a Channel.
+        /// </summary>
+        /// <param name="channel">Name of channel. For eg: .Net Core 5 Dev</param>
+        /// <param name="definitionId">Azure DevOps DefinitionId.</param>
+        /// <param name="minutes">Goal in minutes for a Definition in a Channel.</param>
+        /// <returns>Async task.</returns>
+        public Task<Goal> SetGoalAsync(string channel, int definitionId, int minutes)
+        {
+            var jsonData = new GoalRequestJson(minutes: minutes);
+            return _barClient.Goal.CreateAsync(body : jsonData, channelName : channel, definitionId : definitionId);
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/MaestroApiBarClient.cs]: There is strange whitespace on this line that doesn't match line 427.  
```suggestion
            return _barClient.Goal.CreateAsync(body: jsonData, channelName: channel, definitionId: definitionId);
```

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs (hunk 21 above)
Comment (epananth) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs]: We are taking care of the exceptions in the back-end API part. That is the style we are using. Normally we don't check for null here, that is already taken care of. I will remove the null check here, then it will be better

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/SetGoalOperations.cs (hunk 22 above)
Comment (epananth) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/SetGoalOperations.cs]: according to the document -> https://github.com/dotnet/arcade/blob/master/Documentation/Darc.md , I must print out the goal in minutes.  I will print out the minutes :) thanks Chad!

File: src/Maestro/Client/src/MaestroApiOptions.cs (hunk 17 above)
Comment (epananth) [reply src/Maestro/Client/src/MaestroApiOptions.cs]: changed it

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs (hunk 20 above)
Comment (epananth) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/GetGoalOperation.cs]: added a line between member variables.
